### ISSUE TITLE
AptaTrans: Training code and training schema

### ISSUE THREAD

**Issue by fkiraly** [gc-os-ai/pyaptamer#49]
We should (a) find out how AptaTrans has been trained, and produce a basic, reproducible jupyter notebook, plus data loaders (or data references) for said training.

This is not obvious since the original repository did not incude such code, and it is also not obvious from the paper(s).

Parts:

* investigation on the training procedure and data used, to be reported here
* PR with the training code. This will require #48 to be completed first

**Comment by NennoMP** [comment]
Training of AptaTrans relies on three datasets:
* `bpRNA.db`: [link](https://drive.google.com/file/d/178f_Tb3kgK4sZ23QAhqiK5YOUtkxjyMl/view) (approx. 100MB);
* `protein_ss_keywords.db`: [link](https://drive.google.com/file/d/1E1TIVtppA1L6nu0f8pYQKBAz2ZyBSCMY/view) (approx. 200MB);
* `dataset_li.pickle`: ([link](https://github.com/PNUMLB/AptaTrans/tree/master/data) (approx. 1MB).

The memory estimates assume `.csv` format.

Additionally, there is a `protein_word_freq_3.pickle` file that appears to contain the frequency of all possible triplets of protein letters (e.g., `AAA`, `MCD`, etc.). These are used in the embeddings and I suppose are computed from the `dataset_li.pickle` dataset. Currently, I am still trying to figure out how the frequencies were computed, as I was not able to replicate them.

## Usage
`bpRNA.db` and  `protein_ss_keywords.db` are used to pretrain the encoders for aptamers and proteins respectively, while  `dataset_li.pickle` is used to fine-tune the encoders and the deep neural network for aptamer-protein interaction (API) prediction.

# Origin
Paraphrasing the paper, `bpRNA.db` is a collection of RNA sequences from `bpRNA-1m`, a dataset consisting of 1M RNA sequences from various sources including PBD and Aptamer Base [1]. In the file I counted `102,318` sequences, while in the paper they mention collecting only `79,890` (I suspect a typo related to a previous preprint/version of the paper). The file contains sequences along their secondary structures.

In `protein_ss_keywords.db`, there are `166,136` protein sequences from PDB. The file contains sequences, PIDs, secondary structures and keywords. I guess keywords representing some sort of category for each protein, some examples are _hydrolase_, _hormone_, etc..

Finally, `dataset_li.pickle` is a RNA benchmark from [2]. It contains `2900` total samples already divided into train-test splits used in the paper's experiments. A samples consists of the aptamer and protein sequences, and a (boolean) label determining whether they bing or not.

# File dimensions
I would like to correct myself regarding file sizes. While the fine-tuning dataset is indeed quite small, the other two (dabases) are too big to be put in the repo. We will definitely have them as downloadable with some hosting service.

## References
[1] Cruz-Toledo, Jose, et al. "Aptamer base: a collaborative knowledge base to describe aptamers and SELEX experiments." Database 2012 (2012): bas006.
[2] Li, Bi-Qing, et al. "Prediction of aptamer-target interacting pairs with pseudo-amino acid composition." PLoS One 9.1 (2014): e86729.

**Comment by fkiraly** [comment]
I see - I would suggest `huggingface` or `kaggle` then - maybe let´s check whether they are already there.

**Comment by NennoMP** [comment]
@fkiraly I have temporarily uploaded the two (big) datasets I need on my personal `huggingface`. Should we create a `gc.os` hugging face account and put them there? We could discuss this tomorrow during the weekly.

For reference, you may find the datasets [here](https://huggingface.co/nennomp/datasets).

**Comment by NennoMP** [comment]
@fkiraly Nice. I moved ownership of the datasets to `gcos`.
